### PR TITLE
HIVE-27129: Add enhanced support for Hive Metastore Client http support

### PR DESCRIPTION
**Pull request by junlinzeng-db**
### What changes were proposed in this pull request?
This PR targets to improve the usability of the http hive metastore client, by adding support for custom headers and also, supporting default trust store for https.

Add a new hive conf to allow users to inject custom headers
Do not force the headers to be http when the useSSL is turned off, so that if the users are able to use https if they want, w/o a custom truststore.

### Why are the changes needed?
Improved user experience for http mode in HMS client.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
ut for the http headers

### CODE REVIEW

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
Comment (vihangk1) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:38]: We generally don't use wild-card imports. Can you please revert this part and explicitly add the necessary imports (You may have to change your IDE settings so that it doesn't do this for automatically).

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
Comment (vihangk1) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:622]: Is this log needed? If not, let's remove it or change it to debug.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
Comment (vihangk1) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:673]: Please annotate this with `@VisibleForTesting`

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
Comment (vihangk1) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:682]: nit, unintentional new line?

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
Comment (vihangk1) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:686]: You may want set up your IDE to change the it's default formatting style. There are instructions in https://cwiki.apache.org/confluence/display/Hive/HowToContribute#HowToContribute-CodingConventions

Generally the `@Override` is in a separate line above `public void process` as per the convention.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
Comment (vihangk1) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:705]: same here.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetastoreHttpHeaders.java
Comment (vihangk1) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetastoreHttpHeaders.java:40]: avoid wildcard imports.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetastoreHttpHeaders.java
Comment (vihangk1) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetastoreHttpHeaders.java:42]: Please reformat this file once you have your IDE setup with the formatter above to fix code-style issues.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetastoreHttpHeaders.java
Comment (vihangk1) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetastoreHttpHeaders.java:78]: This method can be in a `BeforeClass` I guess. Also, you probably need a `AfterClass` to shutdown the metastore that you started just to make sure that you clean up after the test completes.

**Review by vihangk1 — COMMENTED** [review]
Overall the change looks good. I left some comments below.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
Comment (junlinzeng-db) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:38]: yeah it was my IDE. fixed

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
Comment (junlinzeng-db) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:673]: do I still need this to be protocted?

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetastoreHttpHeaders.java
Comment (junlinzeng-db) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetastoreHttpHeaders.java:78]: I was mostly copying https://github.com/apache/hive/blob/cd53e0a4d9b99f825beb7a7f5c9acc1b3e822edd/standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestRemoteHiveMetaStoreIpAddress.java#L42, but sure I can make it happen

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
Comment (vihangk-db) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:689]: keeping it as protected doesn't harm. I think this should be okay.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java
Comment (vihangk-db) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java:1591]: nit, please remove the extra line here.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetastoreHttpHeaders.java
Comment (vihangk-db) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetastoreHttpHeaders.java:81]: Ah,  I see. That's interesting. In that case I think it is okay to follow existing pattern.

**Review by vihangk-db — COMMENTED** [review]
Just a small nit about the new line in MetastoreConf. Once that is addressed, I am merge this in. Thanks for addressing the comments.

**Review by vihangk1 — APPROVED** [review]
Will wait for the CI to come back and then merge it. Thanks @junlinzeng-db!

**Comment by vihangk1** [comment]
@junlinzeng-db Jfyi, I fixed the description and the summary title of the PR as the convention.

**Pull request merged**
